Undo/COUndoTrackStore.m

```diff
@@ -283,55 +402,180 @@ - (id) deserialize: (NSData *)data
 {
-	[_db executeUpdate: @"DELETE FROM commands WHERE uuid = ?", [aUUID dataValue]];
+	assert(dispatch_get_current_queue() != _queue);
+	
+	dispatch_sync(_queue, ^() {
+		[_db executeUpdate: @"DELETE FROM commands WHERE uuid = ?", [aUUID dataValue]];
+	});
 }
 
 - (BOOL) string: (NSString *)aString matchesGlobPattern: (NSString *)aPattern
 {
-	return [_db boolForQuery: @"SELECT 1 WHERE ? GLOB ?", aString, aPattern];
+	assert(dispatch_get_current_queue() != _queue);
+	__block BOOL result = NO;
+
+	dispatch_sync(_queue, ^() {
+		result = [_db boolForQuery: @"SELECT 1 WHERE ? GLOB ?", aString, aPattern];
+	});
+	
+	return result;
+}
+
+- (void)markCommandsAsDeletedForUUIDs: (NSArray *)UUIDs
+{
+	assert(dispatch_get_current_queue() != _queue);
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+	
+	__block NSArray *compactedTrackNames = nil;
+
+	// This can be run in background
+	dispatch_sync(_queue, ^() {
+		[_db beginTransaction];
+
+		for (ETUUID *UUID in UUIDs)
+		{
+			[_db executeUpdate: @"UPDATE commands SET deleted = 1 WHERE uuid = ?", [UUID dataValue]];
+		}
+
+		compactedTrackNames =
+			[_db arrayForQuery: @"SELECT DISTINCT trackname FROM commands WHERE deleted = 1 "];
+
+		[_db commit];
+	});
+
+	/* This must be run in the main thread:
+	   - no one must access _modifiedTrackStateForTrackName at the same time
+	   - we must post the commit notification immediately
+	   
+	   If we post it with some delay, someone could touch 
+	   _modifiedTrackStateForTrackName and overwrite COUndoTrackState.compacted. */
+	dispatch_sync_now(dispatch_get_main_queue(), ^() {
+		dispatch_sync(_queue, ^() {
+			[_db beginTransaction];
+
+			for (NSString *trackName in compactedTrackNames)
+			{
+				if (_modifiedTrackStateForTrackName[trackName] == nil)
+				{
+					_modifiedTrackStateForTrackName[trackName] = [self stateForTrackNameInCurrentQueue: trackName];
+				}
+				((COUndoTrackState *)_modifiedTrackStateForTrackName[trackName]).compacted = YES;
+			}
+
+			[_db commit];
+		});
+
+		[self postCommitNotifications];
+	});
+
+	dispatch_semaphore_signal(_transactionLock);
+}
+
+- (void)finalizeDeletions
+{
+	assert(dispatch_get_current_queue() != _queue);
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
+	dispatch_sync(_queue, ^() {
+		[_db executeUpdate: @"DELETE FROM commands WHERE deleted = 1"];
+	});
+	dispatch_semaphore_signal(_transactionLock);
+}
+
+/**
+ * We run vacuum with dispatch_sync() in a queue, so we can sure no other
+ * threads access the database when we invoke -vacuum in a background thread.
+ */
+- (BOOL)vacuum
+{
+    assert(dispatch_get_current_queue() != _queue);
+	__block BOOL success = NO;
+
+	// If there is a transaction underway, wait until it is finished
+	dispatch_semaphore_wait(_transactionLock, DISPATCH_TIME_FOREVER);
```

Comment:
I don't understand why _transactionLock is needed (once the begin/commit transaction calls are moved to _queue as I suggested above).

Just to confirm, the addition of `_queue` is just for future use and doesn't actually provide any concurrency right now, correct?

Once all DB access is moved to `_queue`, it should be safe to make -vacuum use `dispatch_async` instead of `dispatch_sync`. This may not actually help because the next thing to call `dispatch_sync` will block until the vacuum finishes.